x/feedistribution/keeper/allocation.go

```diff
@@ -0,0 +1,148 @@
+			continue
+		}
+		validatorDetail, found := k.StakingKeeper.ValidatorByConsAddrForChainID(
+			ctx, sdk.GetConsAddress(pk), avstypes.ChainIDWithoutRevision(ctx.ChainID()),
+		)
+		if !found {
+			ctx.Logger().Error("Operator address not found; skipping", "consAddress", sdk.GetConsAddress(pk), "i", i)
+			continue
+		}
+		powerFraction := math.LegacyNewDec(val.Power).QuoTruncate(math.LegacyNewDec(totalPreviousPower))
+		reward := feeMultiplier.MulDecTruncate(powerFraction)
+		k.AllocateTokensToValidator(ctx, validatorDetail, reward, feePool)
+		remaining = remaining.Sub(reward)
+	}
+
+	// allocate community funding
+	feePool.CommunityPool = feePool.CommunityPool.Add(remaining...)
+	k.SetFeePool(ctx, feePool)
+
+	return nil
+}
+
+// AllocateTokensToValidator allocate tokens to a particular validator,
+// splitting according to commission.
+func (k Keeper) AllocateTokensToValidator(ctx sdk.Context, val stakingtypes.ValidatorI, tokens sdk.DecCoins, feePool *types.FeePool) {
+	logger := k.Logger()
+	rate := val.GetCommission()
+	commission := tokens.MulDec(rate)
+	shared := tokens.Sub(commission)
+	valBz := val.GetOperator()
+
+	// update current commission
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCommission,
+		sdk.NewAttribute(sdk.AttributeKeyAmount, commission.String()),
+		sdk.NewAttribute(types.EventTypeCommission, val.GetOperator().String()),
+	))
+	currentCommission := k.GetValidatorAccumulatedCommission(ctx, valBz)
+	currentCommission.Commission = currentCommission.Commission.Add(commission...)
+	k.SetValidatorAccumulatedCommission(ctx, valBz, currentCommission)
+	// update current rewards, i.e. the rewards to stakers
+	// if the rewards do not exist it's fine, we will just add to zero.
+	// allocate share tokens to all stakers of this operator.
+	k.AllocateTokensToStakers(ctx, val.GetOperator(), shared, feePool)
+
+	// update outstanding rewards
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeRewards,
+		sdk.NewAttribute(sdk.AttributeKeyAmount, commission.String()),
+		sdk.NewAttribute(types.AttributeKeyValidator, val.GetOperator().String()),
+	))
+
+	// ValidatorOutstandingRewards is the rewards of a validator address.
+	outstanding := k.GetValidatorOutstandingRewards(ctx, valBz)
+	outstanding.Rewards = outstanding.Rewards.Add(tokens...)
+	k.SetValidatorOutstandingRewards(ctx, valBz, outstanding)
+	logger.Info("Allocate tokens to validator successfully", "allocated amount is", outstanding.Rewards.String())
+}
+
+func (k Keeper) AllocateTokensToStakers(ctx sdk.Context, operatorAddress sdk.ValAddress, rewardToAllStakers sdk.DecCoins, feePool *types.FeePool) {
+	logger := k.Logger()
+	avsList, err := k.StakingKeeper.GetOptedInAVSForOperator(ctx, operatorAddress.String())
+	if err != nil {
+		ctx.Logger().Error("avs address lists not found; skipping")
+	}
+	stakersPowerMap, curTotoalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(1)
+	for _, avsAddress := range avsList {
```

Comment:
Could we pass the AVS as a parameter to handle this?

Reply:
Seemed no difference